### ISSUE TITLE
Change default pruning algorithm

### ISSUE THREAD

**Issue by himkt** [optuna/optuna#2906]
## Motivation

`HyperbandPruner` was originally introduced in Optuna v1.1.0 and many users enjoy it in optimizations.
For the next major version of Optuna, we consider making it a default pruning algorithm.

## Description

- Compare the performances of MedianPruner and HyperbandPruner on some benchmark tasks (e.g. HPOBench, NASBench on [kurobako](https://github.com/optuna/kurobako). Other benchmark tasks are welcome!)
- Verify the superiority of HyperbandPruner and it changes a default pruner if we confirm that HyperbandPruner outperforms MedianPruner

**Comment by not522** [comment]
I conducted benchmarks for pruning algorithms implemented in Optuna and reviewed them to select the preferred default pruning algorithm. The framework of the benchmarks is the same as that of the samplers. https://github.com/optuna/optuna/issues/2964#issuecomment-1208974962

I conducted two benchmarks.

- Using `TPESampler` with `multivariate=False` and `True`
- Using `RandomSampler`

# TL;DR

The results show that keeping `MedianPruner` as the default is reasonable. Please note that in the benchmarks, I did not tune the parameters of pruning algorithms and used default arguments because these experiments aim to select the default pruner. Therefore, the following results may not reflect the ability of pruning algorithms, and selecting preferred parameters for each problem may improve scores.

# Summary of Benchmark Experiment

## Problems
We have prepared 8 problems, including 4 hpobench problems and 4 basbench problems. For the details, please see the samplers’ benchmark results.

## Pruners
The benchmark consists of the following 6 pruners: `NopPruner`, `MedianPruner`, `PercentilePruner(percentile=25)`, `PercentilePruner(percentile=75)`, `SuccessiveHalvingPruner`, and `HyperbandPruner`. We skipped testing `ThresholdPruner` because it needs threshold parameters and is unsuitable for the default pruner.

# Benchmark result for `TPESampler`

## Summary

We concluded `MedianPruner` and `PercentilePruner(percentile=75)` with multivariate TPESampler are stronger. The size of the budget determines which `percentile` is better.

## Details

One `budget` is defined as the same computational cost of one trial without pruning.

Benchmark results: budget = 25

<details>

| Sampler | Arguments of the sampler | Pruner | Arguments of the pruner | Borda score | First score | Top3 score |
| - | - | - | - | - | - | - |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | NopPruner | {} | 45 | 1 | 3 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | MedianPruner | {} | 80 | 6 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | PercentilePruner | {“percentile”:25} | 61 | 1 | 5 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | PercentilePruner | {“percentile”:75} | 67 | 2 | 6 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | SuccessiveHalvingPruner | {} | 51 | 1 | 4 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | HyperbandPruner | {} | 51 | 0 | 3 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | NopPruner | {} | 8 | 0 | 1 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | MedianPruner | {} | 15 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | PercentilePruner | {“percentile”:25} | 16 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | PercentilePruner | {“percentile”:75} | 22 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | SuccessiveHalvingPruner | {} | 11 | 0 | 1 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | HyperbandPruner | {} | 15 | 0 | 0 |

</details>

Benchmark results: budget = 50

<details>

| Sampler | Arguments of the sampler | Pruner | Arguments of the pruner | Borda score | First score | Top3 score |
| - | - | - | - | - | - | - |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | NopPruner | {} | 73 | 3 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | MedianPruner | {} | 71 | 1 | 7 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | PercentilePruner | {“percentile”:25} | 32 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | PercentilePruner | {“percentile”:75} | 80 | 6 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | SuccessiveHalvingPruner | {} | 37 | 1 | 2 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | HyperbandPruner | {} | 61 | 0 | 5 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | NopPruner | {} | 21 | 0 | 2 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | MedianPruner | {} | 12 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | PercentilePruner | {“percentile”:25} | 2 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | PercentilePruner | {“percentile”:75} | 32 | 0 | 1 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | SuccessiveHalvingPruner | {} | 6 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | HyperbandPruner | {} | 19 | 0 | 0 |

</details>

Benchmark results: budget = 75

<details>

| Sampler | Arguments of the sampler | Pruner | Arguments of the pruner | Borda score | First score | Top3 score |
| - | - | - | - | - | - | - |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | NopPruner | {} | 75 | 3 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | MedianPruner | {} | 69 | 0 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | PercentilePruner | {“percentile”:25} | 26 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | PercentilePruner | {“percentile”:75} | 83 | 7 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | SuccessiveHalvingPruner | {} | 28 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | HyperbandPruner | {} | 63 | 0 | 6 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | NopPruner | {} | 29 | 0 | 2 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | MedianPruner | {} | 13 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | PercentilePruner | {“percentile”:25} | 3 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | PercentilePruner | {“percentile”:75} | 42 | 0 | 2 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | SuccessiveHalvingPruner | {} | 1 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | HyperbandPruner | {} | 18 | 0 | 1 |

</details>

Benchmark results: budget = 100

<details>

| Sampler | Arguments of the sampler | Pruner | Arguments of the pruner | Borda score | First score | Top3 score |
| - | - | - | - | - | - | - |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | NopPruner | {} | 78 | 3 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | MedianPruner | {} | 63 | 0 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | PercentilePruner | {“percentile”:25} | 19 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | PercentilePruner | {“percentile”:75} | 83 | 7 | 8 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | SuccessiveHalvingPruner | {} | 29 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:true,“constant_liar”:false} | HyperbandPruner | {} | 63 | 0 | 7 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | NopPruner | {} | 41 | 0 | 3 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | MedianPruner | {} | 13 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | PercentilePruner | {“percentile”:25} | 2 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | PercentilePruner | {“percentile”:75} | 47 | 0 | 3 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | SuccessiveHalvingPruner | {} | 0 | 0 | 0 |
| TPESampler | {“n_startup_trials”:10,“n_ei_candidates”:24,“multivariate”:false,“constant_liar”:false} | HyperbandPruner | {} | 35 | 0 | 2 |

</details>

# Benchmark result for `RandomSampler`

## Summary

`MedianPruner` and `SuccessiveHalvingPruner` are stronger.

## Details

Benchmark results: budget = 25

<details>

| Sampler | Arguments of the sampler | Pruner | Arguments of the pruner | Borda score | First score | Top3 score |
| - | - | - | - | - | - | - |
| RandomSampler | {} | NopPruner | {} | 0 | 0 | 5 |
| RandomSampler | {} | MedianPruner | {} | 27 | 5 | 8 |
| RandomSampler | {} | PercentilePruner | {"percentile":25} | 22 | 4 | 8 |
| RandomSampler | {} | PercentilePruner | {"percentile":75} | 10 | 1 | 5 |
| RandomSampler | {} | SuccessiveHalvingPruner | {} | 22 | 5 | 8 |
| RandomSampler | {} | HyperbandPruner | {} | 12 | 1 | 7 |

</details>

Benchmark results: budget = 50

<details>

| Sampler | Arguments of the sampler | Pruner | Arguments of the pruner | Borda score | First score | Top3 score |
| - | - | - | - | - | - | - |
| RandomSampler | {} | NopPruner | {} | 2 | 0 | 4 |
| RandomSampler | {} | MedianPruner | {} | 26 | 3 | 8 |
| RandomSampler | {} | PercentilePruner | {"percentile":25} | 19 | 3 | 8 |
| RandomSampler | {} | PercentilePruner | {"percentile":75} | 13 | 1 | 8 |
| RandomSampler | {} | SuccessiveHalvingPruner | {} | 23 | 5 | 8 |
| RandomSampler | {} | HyperbandPruner | {} | 11 | 1 | 8 |

</details>

Benchmark results: budget = 75

<details>

| Sampler | Arguments of the sampler | Pruner | Arguments of the pruner | Borda score | First score | Top3 score |
| - | - | - | - | - | - | - |
| RandomSampler | {} | NopPruner | {} | 1 | 0 | 4 |
| RandomSampler | {} | MedianPruner | {} | 30 | 5 | 8 |
| RandomSampler | {} | PercentilePruner | {"percentile":25} | 15 | 2 | 6 |
| RandomSampler | {} | PercentilePruner | {"percentile":75} | 16 | 1 | 8 |
| RandomSampler | {} | SuccessiveHalvingPruner | {} | 22 | 4 | 7 |
| RandomSampler | {} | HyperbandPruner | {} | 12 | 0 | 7 |

</details>

Benchmark results: budget = 100

<details>

| Sampler | Arguments of the sampler | Pruner | Arguments of the pruner | Borda score | First score | Top3 score |
| - | - | - | - | - | - | - |
| RandomSampler | {} | NopPruner | {} | 1 | 0 | 2 |
| RandomSampler | {} | MedianPruner | {} | 28 | 4 | 8 |
| RandomSampler | {} | PercentilePruner | {"percentile":25} | 14 | 1 | 8 |
| RandomSampler | {} | PercentilePruner | {"percentile":75} | 15 | 2 | 7 |
| RandomSampler | {} | SuccessiveHalvingPruner | {} | 27 | 4 | 8 |
| RandomSampler | {} | HyperbandPruner | {} | 13 | 1 | 7 |

</details>

**Comment by not522** [comment]
Let me close this issue since we have concluded keeping `MedianPruner` as the default.
